lib/resumption/resumable.go

```diff
@@ -0,0 +1,339 @@
+)
+
+const maxFrameSize = 128 * 1024
+
+// runResumeV1Unlocking runs the symmetric resumption v1 protocol for r, using
+// nc as the underlying transport. The previous attached transport, if any, will
+// be detached immediately. firstConn signifies that the connection has not been
+// used, and the initial handshake will be assumed to be 0 for both sides. The
+// connection lock is assumed to be held when entering the function, since the
+// correct behavior of firstConn requires no possible external interference
+// before the attach point is reached; the lock will be not held when the
+// function returns.
+func runResumeV1Unlocking(r *Conn, nc net.Conn, firstConn bool) error {
+	defer nc.Close()
+
+	if !firstConn {
+		for !r.remoteClosed && r.requestDetach != nil {
+			r.requestDetach()
+			r.cond.Wait()
+		}
+
+		if r.remoteClosed {
+			r.mu.Unlock()
+
+			return trace.ConnectionProblem(errBrokenPipe, "attempting to resume a connection already closed by the peer")
+		}
+	} else if r.requestDetach != nil || r.remoteClosed || r.localClosed || r.receiveBuffer.end > 0 || r.sendBuffer.start > 0 {
+		r.mu.Unlock()
+		panic("firstConn for resume V1 is not actually unused")
+	}
+
+	var stopRequested atomic.Bool
+
+	requestStop := func() {
+		if !stopRequested.Swap(true) {
+			r.cond.Broadcast()
+		}
+	}
+	r.requestDetach = func() {
+		nc.Close()
+		requestStop()
+	}
+	r.cond.Broadcast()
+
+	defer func() {
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		r.requestDetach = nil
+		r.cond.Broadcast()
+	}()
+
+	sentPosition := r.receiveBuffer.end
+	r.mu.Unlock()
+
+	ncReader, ok := nc.(byteReaderReader)
+	if !ok {
+		ncReader = bufio.NewReader(nc)
+	}
+
+	var peerPosition uint64
+
+	handshakeWatchdog := time.AfterFunc(handshakeTimeout, func() { nc.Close() })
+	defer handshakeWatchdog.Stop()
+
+	if !firstConn {
+		errC := make(chan error, 1)
+		go func() {
+			_, err := nc.Write(binary.AppendUvarint(nil, sentPosition))
+			errC <- err
+		}()
+		var err error
+		peerPosition, err = binary.ReadUvarint(ncReader)
+		if err != nil {
+			return trace.Wrap(err, "reading peer receive position")
+		}
+		err = <-errC
+		if err != nil {
+			return trace.Wrap(err, "writing receive position")
+		}
+	}
+
+	r.mu.Lock()
+	if minPos, maxPos := r.sendBuffer.start, r.sendBuffer.end; peerPosition < minPos || maxPos < peerPosition {
+		// incompatible resume position, mark as remotely closed since we can't
+		// ever continue from this; this also includes receiving an errorTag
+		// (since that's too big of a position to reach legitimately)
+		r.remoteClosed = true
+		r.cond.Broadcast()
+		r.mu.Unlock()
+
+		_, _ = nc.Write([]byte(errorTagUvarint))
+		return trace.BadParameter("got incompatible resume position (%v, expected %v to %v)", peerPosition, minPos, maxPos)
+	}
+	if r.sendBuffer.start != peerPosition {
+		r.sendBuffer.advance(peerPosition - r.sendBuffer.start)
+		r.cond.Broadcast()
+	}
+	r.mu.Unlock()
+
+	handshakeWatchdog.Stop()
```

Comment:
if it's not too onerous, can you refactor this into a `handshake` function, just to make it easier to understand what it's doing?